crates/chia-datalayer/src/lib.rs

```diff
@@ -0,0 +1,1259 @@
+//     }
+// }
+
+fn internal_hash(left_hash: Hash, right_hash: Hash) -> Hash {
+    let mut hasher = Sha256::new();
+    hasher.update(b"\x02");
+    hasher.update(left_hash);
+    hasher.update(right_hash);
+
+    hasher.finalize()
+}
+
+pub struct DotLines {
+    nodes: Vec<String>,
+    connections: Vec<String>,
+    pair_boxes: Vec<String>,
+}
+
+impl Default for DotLines {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DotLines {
+    pub fn new() -> Self {
+        Self {
+            nodes: vec![],
+            connections: vec![],
+            pair_boxes: vec![],
+        }
+    }
+
+    pub fn push(&mut self, mut other: DotLines) {
+        self.nodes.append(&mut other.nodes);
+        self.connections.append(&mut other.connections);
+        self.pair_boxes.append(&mut other.pair_boxes);
+    }
+
+    pub fn dump(&mut self) -> String {
+        // TODO: consuming itself, secretly
+        let mut result = vec!["digraph {".to_string()];
+        result.append(&mut self.nodes);
+        result.append(&mut self.connections);
+        result.append(&mut self.pair_boxes);
+        result.push("}".to_string());
+
+        result.join("\n")
+    }
+}
+
+const NULL_PARENT: TreeIndex = 0xffff_ffffu32;
+
+#[derive(Debug, PartialEq)]
+pub struct NodeMetadata {
+    pub node_type: NodeType,
+    pub dirty: bool,
+}
+
+impl NodeMetadata {
+    pub fn from_bytes(blob: [u8; METADATA_SIZE]) -> Result<Self, String> {
+        // TODO: could save 1-2% of tree space by packing (and maybe don't do that)
+        // TODO: identify some useful structured serialization tooling we use
+        Ok(Self {
+            node_type: Self::node_type_from_bytes(blob)?,
+            dirty: Self::dirty_from_bytes(blob)?,
+        })
+    }
+
+    pub fn to_bytes(&self) -> [u8; METADATA_SIZE] {
+        [self.node_type.to_u8(), u8::from(self.dirty)]
+    }
+
+    pub fn node_type_from_bytes(blob: [u8; METADATA_SIZE]) -> Result<NodeType, String> {
+        NodeType::from_u8(blob[0])
+    }
+
+    pub fn dirty_from_bytes(blob: [u8; METADATA_SIZE]) -> Result<bool, String> {
+        match blob[1] {
+            0 => Ok(false),
+            1 => Ok(true),
+            other => Err(format!("invalid dirty value: {other}")),
+        }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Node {
+    parent: Parent,
+    hash: Hash,
+    specific: NodeSpecific,
+    // TODO: kinda feels questionable having it be aware of its own location
+    index: TreeIndex,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum NodeSpecific {
+    Internal { left: TreeIndex, right: TreeIndex },
+    Leaf { key_value: KvId },
+}
+
+impl NodeSpecific {
+    pub fn sibling_index(&self, index: TreeIndex) -> TreeIndex {
+        let NodeSpecific::Internal { right, left } = self else {
+            panic!()
+        };
+
+        match index {
+            x if (x == *right) => *left,
+            x if (x == *left) => *right,
+            _ => panic!(),
```

Comment:
when traversing up you know your index and sometimes you need your sibling.  you don't know whether your sibling is on the left or right until you check which side of your parent you are on.  hence taking the index instead of a side directly.